ccf_transaction_processor/transaction_processor/pdo_tp.cpp

```diff
@@ -0,0 +1,624 @@
+
+using namespace std;
+using namespace ccf;
+using namespace tls;
+
+
+namespace ccfapp
+{
+
+    TPHandler::TPHandler(Store& store):
+        UserHandlerRegistry(store),
+        enclavetable(store.create<string, EnclaveInfo>("encalves")),
+        contracttable(store.create<string, ContractInfo>("contracts")),
+        ccltable(store.create<string, ContractStateInfo>("ccl_updates")),
+        signer(store.create<string, map<string, string>>("signer"))
+    {
+        ledger_signer_local = NULL;
+    }
+
+    string TPHandler::sign_document(const string& document){
+        vector<uint8_t> doc_vector(document.begin(), document.end());
+        auto sign = ledger_signer_local->sign(doc_vector);
+        return b64_from_raw(sign.data(), sign.size());
+    }
+
+    void TPHandler::init_handlers(Store& store){
+
+        UserHandlerRegistry::init_handlers(store);
+
+        //======================================================================================================
+        // ping handler implementation
+        auto ping = [this](Store::Tx& tx, const nlohmann::json& params) {
+            return make_success(true);
+        };
+
+        //======================================================================================================
+        // get_ledger_key implementation
+        auto get_ledger_key = [this](Store::Tx& tx, const nlohmann::json& params) {
+            auto signer_view = tx.get_view(signer);
+            auto signer_global = signer_view->get_globally_committed("signer");
+            if (signer_global.has_value()){
+                try{
+                    auto key_pair = signer_global.value();
+                    if (ledger_signer_local == NULL){
+                        auto privk_pem = tls::Pem(key_pair["privk"]);
+                        ledger_signer_local = make_key_pair(privk_pem, nullb, false);
+                    }
+                    return make_success(Get_Ledger_Key::Out{key_pair["pubk"]});
+                }
+                catch(...){
+                    return make_error(
+                        jsonrpc::StandardErrorCodes::INVALID_PARAMS, "Unable to locate ledger Key");
+                }
+            }
+
+            auto signer_local = signer_view->get("signer");
+            try{
+                if (!signer_local.has_value()){
+                    // create the key, done only the first time this rpc is called
+                    auto curve = CurveImpl::secp256k1_mbedtls;
+                    auto kp = make_key_pair(curve);
+                    auto privk_pem = kp->private_key_pem();
+                    auto pubk_pem = kp->public_key_pem();
+                    map<string, string> key_pair;
+                    key_pair["privk"] = privk_pem.str();
+                    key_pair["pubk"] = pubk_pem.str();
+                    signer_view->put("signer", key_pair);
+                }
+                return make_success("Ledger signing key not globally commited yet. Check back later");
```

Comment:
The function get_ledger_key does two purposes. 

1. Creates the signer the very first time it is called after the instantiation of the service. Specifically creates it locally & schedules for global commit. This call returns as success, with the message that ""Ledger signing key not globally commited yet. Check back later". This call is issued by the sys admin soon after deploying the CCF ledger. The admin can invoke this call multiple times to ensure that the signer has been globally committed. (see key_test.py in CCF tests).

2. For use by PDO clients to get the verifying keys. In this case, since the singer is globally committed, the above scenario never occurs. The client always gets the verifying key since it is globally committed.

Now the real question is can the signer generation be made part of CCF deployment itself, without having to issue an rpc. I don't know, but this current method is simple enough.

Reply:
I'm fine with this function serving two purposes. For one, it think that having to parse the two very different success messages that are returned on an actual "get" versus a pre-global commit query would be kind of clunky. 

On the other hand, why can we assume that the ledger key will be globally committed by the time a client sends a query for the ledger key? I'm not sure we should assume that (malicious) clients will not query the ledger early, since presumably, they would want to obtain the ledger key as soon as they come online. So, for clarity and safety, I think the interface should distinguish between the key generation event and the pre-global commit query. If the key has been generated successfully, the sysadmin should receive a clear success ACK. In that case, it's implied that the key has not yet been committed globally. But trying to obtain they key prior to global commit should be treated as an error since no other functionality _should_ be used until the ledger key has been committed globally.

Reply:
The ledger user only sees one type of success message. it is just the admin that gets the two types. Other option is to create a a separate rpc just to create the new  key, which I feel is unnecessary since this is one time & by the admin.

Yes, you should assume that the  the ledger key will be globally committed by the time a client sends a query for the ledger key. The CCF ledger is expected to be open for use only after "proper deployment", which includes the admin ensuring that the signing key is globally committed.

Reply:
out of band with this PR. I also don't see the value in addressing the issues such as handling  "lazy sys admin". I will clarify in this documentation that the CCF deployer must ensure that signing keys are globally committed.

Reply:
I don't agree. There is definitely a lot of value in preventing unintended security problems caused by Lazy/inexperienced sys admins. This seems like a fairly straight-forward fix. Since PDO isn't being used in production right now, we can probably get away with assuming that the sys admin will do the right thing, but this makes me rather uncomfortable from a security standpoint. Will open an issue for this.

Reply:
I have clarified in the Readme.md that "the CCF deployer must ensure that signing keys are globally committed.". I don't see any security issue here. If the deployer does not guarantee and open up the CCF deployment, a cleint read rpc will not return verifying key  but will simply return the message that "check back later, not globally committed). So, essentially, the client cannot take advantage of the proof of commit until global commit happens.